lib/stateful/animation.dart

```diff
@@ -0,0 +1,288 @@
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+
+class StatefulAnimation extends StatefulWidget {
+  StatefulAnimation({
+    Key key,
+  }): super(key: key);
+
+  @override
+  _StatefulAnimationState createState() => _StatefulAnimationState();
+}
+
+class _StatefulAnimationState extends State<StatefulAnimation> {
+  ValueNotifier<Color> _colorListenable;
+  Timer _timer;
+
+  @override
+  void initState() {
+    super.initState();
+    _colorListenable = ValueNotifier<Color>(Colors.green[100]);
+    int color = 100;
+    _timer = Timer.periodic(const Duration(milliseconds: 900), (Timer timer) {
+      color += 100;
+      if (color > 900)
+        color = 100;
+      _colorListenable.value = Colors.green[color];
+    });
+  }
+
+  @override
+  void dispose() {
+    _timer.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Example(
+      restorationId: 'animation',
+      color: _colorListenable,
+      active: true,
+    );
+  }
+}
+
+class Example extends StatefulWidget {
+  Example({
+    Key key,
+    @required this.restorationId,
+    this.color,
+    this.active,
+  }): assert(restorationId != null),
+      assert(color != null),
+      assert(active != null),
+      super(key: key);
+
+  final String restorationId;
+
+  final ValueListenable<Color> color;
+
+  final bool active;
+
+  @override
+  _ExampleState createState() => _ExampleState();
+}
+
+class _ExampleState extends State<Example> with TickerProviderStateMixin, RestorationMixin {
+  RestorableDuration _duration = RestorableDuration(const Duration(seconds: 5));
+  AnimationController _animation;
+  Color _color;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.color.addListener(_colorChange);
+    _color = widget.color.value;
+  }
+
+  bool _firstTime = true;
+
+  @override
+  void restoreState(RestorationBucket oldBucket) {
+    registerForRestoration(_duration, 'duration');
+    if (_firstTime)
+      _animation = AnimationController(vsync: this);
+    _animation.duration = _duration.value;
+    if (_firstTime && widget.active)
+      _animation.repeat();
+    _firstTime = false;
```

Comment:
What do these lines do?

Reply:
initialize the AnimationController and maintain its duration in the face of state restoration (e.g. hitting the back button on Web after clicking the button to change the duration).